### PR TITLE
Add optimizer state CPU offload

### PR DESCRIPTION
**Pull request by samsja**
## Summary

Adds `CPUOffloadOptimizer` - keeps optimizer states on CPU, moves to GPU only during `optimizer.step()`.

Unlike FSDP's `CPUOffload`, this keeps weights on GPU (no H2D all-gather on forward pass).

## Key insight

With activation checkpointing, activations and optimizer states are never on GPU simultaneously:
- Forward/backward: activations on GPU, optimizer states on CPU
- Optimizer step: optimizer states on GPU, activations freed

Peak memory becomes `max(activations, optimizer_states)` instead of `activations + optimizer_states`.

## Results

**4B model, 4 GPUs, AC enabled:**
- 2 GB saved per GPU (~8% reduction)
- ~2% slower

Best for memory-constrained scenarios where fitting larger model/batch matters more than throughput.

## Usage

```toml
[model]
optim_cpu_offload = true
```

ckpt  resume tests
image

---

> [!NOTE]
> **Medium Risk**
> Touches optimizer stepping, scheduler binding, and checkpoint save/load paths; mistakes could break resume correctness or cause device-transfer/perf issues during training.
> 
> **Overview**
> Adds **optimizer-state CPU offloading** via a new `CPUOffloadOptimizer` wrapper that keeps optimizer states on CPU and transfers them to CUDA only during `step()`.
> 
> Introduces `model.optim_cpu_offload` config (mutually exclusive with `fsdp_cpu_offload`) and wires it into both RL and SFT trainers, with RL additionally disallowing use when `max_concurrent_runs > 1`. Checkpointing and scheduler setup are updated to unwrap/wrap base optimizers so optimizer state save/load works correctly when offloading is enabled.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b6a64f1c47f7c0c6c66f3e19775cfb70126b7783. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/trainer/optim.py [hunk 1]
```diff
@@ -12,12 +13,104 @@
...
+    at the same time: peak memory becomes max(activations, opt_states) instead of sum.
+    """
+
+    def __init__(self, optimizer: Optimizer, pin_memory: bool = True):
+        self.optimizer = optimizer
+        self.pin_memory = pin_memory
+        self._initialized = False
+
+    def _move_states(self, device: str):
+        from torch.distributed.tensor import DTensor
```
Comment (Jackmin801) [inline src/prime_rl/trainer/optim.py]: move this import up?

**Review by mikasenghaas — COMMENTED** [review]
do we wanna deprecate the fsdp offload then?

**Comment by samsja** [comment]
> do we wanna deprecate the fsdp offload then?

I would still keep fsdp offload, its very minor right now and can be usefull for very low gpu ressources

File: src/prime_rl/trainer/optim.py (hunk 1 above)
Comment (samsja) [reply src/prime_rl/trainer/optim.py]: @cursoragent can u move the import up ?

**Review by S1ro1 — APPROVED** [review]
Do you have some plots proving convergence? Ideally w checkpoint restart. In general looks good but skeptical without proof

File: src/prime_rl/trainer/optim.py [hunk 2]
```diff
@@ -12,12 +15,110 @@
...
+                    local_tensor = v._local_tensor
+                    if device == "cpu":
+                        non_blocking = not self.pin_memory
+                        new_local = local_tensor.to("cpu", non_blocking=non_blocking)
+                        if self.pin_memory and not new_local.is_pinned():
+                            new_local = new_local.pin_memory()
+                    else:
+                        target_device = torch.device("cuda")
+                        if isinstance(p, DTensor):
+                            target_device = p._local_tensor.device
```
Comment (S1ro1) [inline src/prime_rl/trainer/optim.py]: Why are we overriding the target device here? Shouldn't we send to a device as specified by the function argument?

File: src/prime_rl/trainer/optim.py [hunk 3]
```diff
@@ -12,12 +15,110 @@
...
+                    state[k] = new_dtensor
+                elif isinstance(v, torch.Tensor):
+                    if device == "cpu":
+                        non_blocking = not self.pin_memory
+                        cpu_tensor = v.to("cpu", non_blocking=non_blocking)
+                        if self.pin_memory and not cpu_tensor.is_pinned():
+                            cpu_tensor = cpu_tensor.pin_memory()
+                        state[k] = cpu_tensor
+                    else:
+                        target_device = p.device if hasattr(p, "device") else "cuda"
```
Comment (S1ro1) [inline src/prime_rl/trainer/optim.py]: Likewise here

**Comment by samsja** [comment]
Re: target_device override question - The `device` arg is a direction flag, not a specific target. When moving back to GPU, we need to match each state tensor to its parameter's device (cuda:0, cuda:1, etc.) for multi-GPU FSDP. Using just `"cuda"` would default to the current device, which could be wrong.

File: src/prime_rl/trainer/scheduler.py [hunk 4]
```diff
@@ -1,14 +1,19 @@
...
 from prime_rl.utils.logger import get_logger
 
-if TYPE_CHECKING:
-    from prime_rl.trainer.optim import MultiLoRAOptimizer
+
+def _get_base_optimizer(optimizer: Optimizer | CPUOffloadOptimizer) -> Optimizer:
+    """Extract the base optimizer from a potentially wrapped optimizer."""
+    if isinstance(optimizer, CPUOffloadOptimizer):
+        return optimizer.optimizer
+    return optimizer
```
[the remark this answers is not in the document]
Comment (S1ro1) [reply src/prime_rl/trainer/scheduler.py:16]: What is being suggested here is good, we should use the `._get_base_optimizer` or just match them

**Review by S1ro1 — APPROVED** [review]
lgtm

File: src/prime_rl/trainer/config.py [hunk 5]
```diff
@@ -167,6 +167,13 @@ class ModelConfig(BaseConfig):
         ),
     ] = False
 
+    optim_cpu_offload: Annotated[
+        bool,
+        Field(
+            description="Whether to enable optimizer state CPU offloading. Unlike fsdp_cpu_offload, this only moves optimizer states (momentum, variance) to CPU, keeping weights on GPU. This avoids the H2D all-gather overhead while still saving GPU memory.",
+        ),
+    ] = False
```
[the remark this answers is not in the document]
Comment (S1ro1) [reply src/prime_rl/trainer/config.py:175]: Also this lol???

[the remark this answers is not in the document]
Comment (samsja) [reply src/prime_rl/trainer/config.py:175]: NAH WHAT ?????????????????????
